### PR TITLE
[JENKINS-46725] stop relying on jnr-posix

### PR DESCRIPTION
**Pull request by ndeloof**
See [JENKINS-46725](https://issues.jenkins-ci.org/browse/JENKINS-46725).

### Proposed changelog entries

* Removed jnr-posix as dependency, relying on plain JDK APIs for filesystem access

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)

### Desired reviewers

@jenkinsci/java10-support @oleg-nenashev 

I guess this approach will be controversial, as it **removes** stuff from jenkins-core. Way too hack-ish anyway, highly improbable jnr-posix library will survive to jigsaw.

### CODE REVIEW

**Comment by daniel-beck** [comment]
Javadoc errors due to `{@link POSIX`.

File: core/pom.xml [hunk 1]
```diff
@@ -106,11 +106,6 @@ THE SOFTWARE.
       <artifactId>jna-posix</artifactId>
       <version>1.0.3-jenkins-1</version>
     </dependency>
-    <dependency>
-      <groupId>com.github.jnr</groupId>
-      <artifactId>jnr-posix</artifactId>
```
Comment (oleg-nenashev) [inline core/pom.xml]: There are usages outside the core: https://github.com/jenkinsci/ssh-agent-plugin/blob/fd42f5e7aeb3772e1baf264cc1b39761dcda79a0/src/main/java/com/cloudbees/jenkins/plugins/sshagent/jna/AgentServer.java

File: core/src/main/java/hudson/os/PosixAPI.java [hunk 2]
```diff
@@ -5,46 +5,25 @@
...
-                    try {
-                        return org.jruby.ext.posix.POSIX.ERRORS.valueOf(error.name());
-                    } catch (IllegalArgumentException x) {
-                        return org.jruby.ext.posix.POSIX.ERRORS.EIO; // PosixException.message has real error anyway
-                    }
-                }
-            }, true);
-        }
-        return posix;
+    public static synchronized Object jnr() {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/os/PosixAPI.java]: There are usages outside the core: https://github.com/search?q=org%3Ajenkinsci+%22PosixAPI%22&type=Code

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
There are usages outside the core:

* https://github.com/search?q=org%3Ajenkinsci+%22PosixAPI%22&type=Code
* https://github.com/search?l=Java&q=org%3Ajenkinsci+%22jnr.posix%22&type=Code

Major plugins will break, including SSH Agent and PAM Auth

My proposal would be:

* Detach `hudson.os.PosixAPI` to a plugin (E.g. JNR POSIX API Plugin). It will retain compatibility for anybody who wants to use it
* Remove the NATIVE_CHMOD mode from the core 
* OR: offer a "PosixAPIProvider" extension point to generalize it and then implement it in the plugin

Removal of Native POSIX option looks feasible. https://github.com/search?q=%22.useNativeChmodAndMode%22&type=Code shows there is no usages in public repos. 

Offering an extension point Maybe reasonable if we want to try experimenting with Project Panama later.

**Comment by oleg-nenashev** [comment]
@jenkinsci/java10-support

**Comment by oleg-nenashev** [comment]
Anyway, I like the idea of removing JNR depednency from the core

**Comment by ndeloof** [comment]
I can prepare a jnr-posix-plugin to support backward compatibility, but will need to also fix those consuming plugins, as they won't be able to run on future JVMs.

**Comment by oleg-nenashev** [comment]
@ndeloof Yes, it makes sense to create follow-up tickets for them. Fixes will be reasonable for sure (if possible).

**Comment by dwnusbaum** [comment]
+1 on the idea. It's great that we already cleaned up the main uses inside of core as part of [JENKINS-36088](https://issues.jenkins-ci.org/browse/JENKINS-36088) so detaching it is straightforward. However, I think we'd want to detach the `jna-posix` dependency at the same time, and I think that still has some uses that are not easily replaceable.

> Remove the NATIVE_CHMOD mode from the core

The `useNativeChmodAndMode` option is relatively new (added in #3135 in Jenkins 2.93), and was only added in case there were any regressions when switching to the Java implementation by default. I am not aware of any regressions caused by the Java implementation, so my guess is that the option is unused, and can be removed without issue.

**Comment by oleg-nenashev** [comment]
We can do the extension point later if needed (I can create a follow-up ticket for Panama project), so I am fine with just removing the feature for now to get it going. 

CC @Typz who participated in the original change request

**Comment by ndeloof** [comment]
see https://github.com/ndeloof/jnr-posix-api-plugin

**Comment by ndeloof** [comment]
main limitation (which a unit test reveal) is we will miss support to set setuid/setgid bits on file.
not sure this is actually used somewhere (unfortunately harder to search on github) and if this is desirable for jenkins to support this

**Comment by dwnusbaum** [comment]
> main limitation (which a unit test reveal) is we will miss support to set setuid/setgid bits on file

The actual behavior that `FilePathTest#copyToWithPermissionSpecialPermissions` is checking (that `FilePath#copyToWithPermission` uses `Files.copy` with `StandardCopyOption.COPY_ATTRIBUTES` where possible to copy special attributes even if those attributes cannot be accessed from Java) will still work fine once JNR is removed, we just won't have a way to set the `setuid/setgid` bits it to be able to test it unless we do something like run `chmod` in a separate process on the file directly.

**Comment by ndeloof** [comment]
jnr-posix-plugin & jna-posix-plugin released

**Review by dwnusbaum — CHANGES_REQUESTED** [review]
There are quite a few test failures in the CI build. At least some of them are because HudsonTestCase uses PosixAPI in the test harness, see [here](https://github.com/jenkinsci/jenkins-test-harness/blob/53f82026d9d143535dd693c2617fd449c3a49741/src/main/java/org/jvnet/hudson/test/HudsonTestCase.java#L1776-L1777). I don't know if there is a pure Java replacement for that code, but we need to remove PosixAPI uses from the harness in some way before it can be detached.

**Comment by oleg-nenashev** [comment]
@ndeloof are you still interested in it? @dwnusbaum has updated JNR POSIX in another pull-request, so we have no IllegalReflectiveAccess issues with it anymore (AFAICT)

**Comment by dwnusbaum** [comment]
> so we have no IllegalReflectiveAccess issues with it anymore

Maybe. I know we don't get them consistently at startup anymore, but if we use certain code paths we could still get them. JNR still uses illegal reflective access (and that probably cannot be changed), it just moved to using lazy initialization rather than static initializers for some of those calls to avoid warnings if you don't use those code paths. See [this Jira comment](https://issues.jenkins-ci.org/browse/JENKINS-46619?focusedCommentId=356635&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-356635) for more details.

**Comment by philippe-granet** [comment]
@dwnusbaum 
> The `useNativeChmodAndMode` option is relatively new (added in #3135 in Jenkins 2.93), and was only added in case there were any regressions when switching to the Java implementation by default. I am not aware of any regressions caused by the Java implementation, so my guess is that the option is unused, and can be removed without issue.

I think there is some problem with the java implementation when chmod is used to launch an external shell script:
https://issues.jenkins-ci.org/browse/JENKINS-53387
https://issues.jenkins-ci.org/browse/JENKINS-48258

I have this problem on my project (several times per week) with `sh` step of `durable-task-plugin` and with `ssh-agent-plugin`.
These 2 plugings generate a shell script , the use FilePath.chmod to make it executable, and then launch the shell script:
https://github.com/jenkinsci/durable-task-plugin/blob/master/src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java  
https://github.com/jenkinsci/ssh-agent-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/sshagent/exec/ExecRemoteAgent.java

It seems to be a JDK bug, see this comment: https://issues.jenkins-ci.org/browse/JENKINS-48258?focusedCommentId=324590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-324590
> The issue here appears to be caused by a Java bug, https://bugs.openjdk.java.net/browse/JDK-8068370. The issue is possible when multiple threads open a file for writing, close it, and then execute them (each thread using its own file). Even if all files are closed "properly", due to how file handles work around fork/exec, a child process in one thread may inherit the handle to anther thread's open file, and thus break that thread's later subprocess call.

**Comment by dwnusbaum** [comment]
@philippe-granet Probably better to have this discussion over in a Jira ticket, but does adding `-D hudson.Util.useNativeChmodAndMode=true` as a system property to Jenkins fix the issue? From a quick read of the linked JDK bug and tickets, what you are describing looks like an issue with writing to files and then executing them in a subprocess, and it seems like even the native implementation of `FilePath#chmod` would have the same issue.

**Comment by batmat** [comment]
Given the status here  (mainly needs-fix one year ago, + the unresolved conflicts), we agreed with @daniel-beck to close this. This is unfortunate, but obviously this does not mean it's not possible to reopen.

Please Nicolas or anyone feel free to comment/ask for reopening anytime if you feel like it.

Thanks again for the work!

**Pull request closed without merging**